HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+    ----------
+    shocks : np.array
+        Length-3 array with the stochastic shocks that get realized between the
+        end of the current period and the start of next period. Their order is
+        (0) permanent income shock, (1) transitory income shock, (2) risky
+        asset return.
+    nNrm : float
+        End-of-period risky asset balances.
+    Share : float
+        End-of-period income deduction share.
+    PermGroFac : float
+        Permanent income growth factor.
+
+    Returns
+    -------
+    n_nrm_tp1 : float
+        Next-period normalized risky balance.
+
+    """
+
+    # Extract shocks
+    perm_shk = shocks[0]
+    tran_shk = shocks[1]
+    R_risky = shocks[2]
+
+    n_nrm_tp1 = R_risky * nNrm / (perm_shk * PermGroFac) + Share * tran_shk
+
+    return n_nrm_tp1
+
+
+def end_of_period_derivs(
+    shocks,
+    a,
+    nTil,
+    s,
+    dvdm_next,
+    dvdn_next,
+    dvds_next,
+    CRRA,
+    PermGroFac,
+    Rfree,
+    DiscFac,
+    LivPrb,
+    v_next=None,
+):
+    """
+    While solving his problem in a given period, the agent must estimate the
+    expected continuation value of ending the current period in a given state.
+    Most importantly, he must know the derivatives of that function.
+    This method is a step in computing these expected values. It computes the
+    end-of-period derivatives (and optionally the value) of the continuation
+    function, conditional on shocks. This is so that the expectations can be
+    calculated later by integrating over shocks.
+
+    Parameters
+    ----------
+    shocks : np.array
+        Length-3 array with the stochastic shocks that get realized between the
+        end of the current period and the start of next period. Their order is
+        (0) permanent income shock, (1) transitory income shock, (2) risky
+        asset return.
+    a : float
+        end-of-period risk-free assets.
+    nTil : float
+        end-of-period risky assets.
+    s : float
+        end-of-period income deduction share.
+    dvdm_next : 3D function
+        Next-period's marginal value of riskless assets.
+    dvdn_next : 3D function
+        Next-period's marginal value of risky assets.
+    dvds_next : 3D function
+        Next-period's marginal value of income-deduction share.
+    CRRA : float
+        Coefficient of relative risk aversion.
+    PermGroFac : float
+        Permanent income deterministic growth factor.
+    Rfree : float
+        Risk-free return factor.
+    DiscFac : float
+        Time-preference discount factor.
+    LivPrb : float
+        Survival probability.
+    v_next : 3D function, optional
+        Next-period's value function. The default is None.
+
+    Returns
+    -------
+    np.array
+        Array with end-of-period-value derivatives conditional on next
+        period's shocks. Order
+        (0) riskless assets, (1) risky assets, (2) income deduction share
+        Optionally, the level of end-of-period value is added in position (3).
```

Comment:
This seems like an awfully contrived data structure to me.
Is there a reason why these are getting stacked together rather than returned as separate arrays?

Reply:
Yes, there's a reason. I recognize the ugliness. It is built like that to be called by `calc_expectation`, let me explain.

At this stage I need to compute the expectation of multiple functions of next-period's states, at every post-state gridpoint. Something like `E[{f(m(a,shocks)), g(m(a, shocks))}]` taken over `shocks`.

You are right that I could compute, for instance, `{E[f(m(a,shocks))], E[g(m(a, shocks))]}` which would be clearer, but this would duplicate the amount of times that I must compute `m(a,shocks)`. Since this is the part of the solution that takes the longest, I opted for speed over organization.

There might be a dominant approach that I have not thought of, like first finding the distribution of `m|a` and then taking the expectations using that. The tricky bit is that there would be a distribution of `m` for every `a` and so there might be some wrangling of distribution objects that might be even uglier and slower than the current solution.